services/clusters-keeper/src/simcore_service_clusters_keeper/modules/clusters_management_core.py

```diff
@@ -21,16 +25,44 @@
 _logger = logging.getLogger(__name__)
 
 
-def _get_instance_last_heartbeat(instance: EC2InstanceData) -> datetime.datetime:
-    if last_heartbeat := instance.tags.get(HEARTBEAT_TAG_KEY, None):
+def _get_instance_last_heartbeat(instance: EC2InstanceData) -> datetime.datetime | None:
+    if last_heartbeat := instance.tags.get(
+        HEARTBEAT_TAG_KEY,
+    ):
         last_heartbeat_time: datetime.datetime = arrow.get(last_heartbeat).datetime
         return last_heartbeat_time
-    launch_time: datetime.datetime = instance.launch_time
-    return launch_time
+
+    return None
+
+
+_USER_ID_TAG_KEY: Final[AWSTagKey] = parse_obj_as(AWSTagKey, "user_id")
+_WALLET_ID_TAG_KEY: Final[AWSTagKey] = parse_obj_as(AWSTagKey, "wallet_id")
+
+
+async def _get_all_associated_worker_instances(
+    app: FastAPI,
+    primary_instances: Iterable[EC2InstanceData],
+) -> list[EC2InstanceData]:
+    worker_instances = []
+    for instance in primary_instances:
+        assert "user_id" in instance.tags  # nosec
+        user_id = UserID(instance.tags[_USER_ID_TAG_KEY])
+        assert "wallet_id" in instance.tags  # nosec
+        # NOTE: wallet_id can be None
+        wallet_id = (
+            WalletID(instance.tags[_WALLET_ID_TAG_KEY])
+            if instance.tags[_WALLET_ID_TAG_KEY] != "None"
```

Comment:
so this is because tags can only be strings, right? Therefore it was previously assigned as `instance.tags[_WALLET_ID_TAG_KEY] = str(None) `